go/vt/srvtopo/query.go

```diff
@@ -0,0 +1,162 @@
+	if !cacheValid && staleOK {
+		// Only allow stale results for a bounded period
+		cacheValid = entry.value != nil && (time.Since(entry.insertionTime) < (q.cacheTTL + 2*q.cacheRefresh))
+	}
+	shouldRefresh := time.Since(entry.lastQueryTime) > q.cacheRefresh
+
+	// If it is not time to check again, then return either the cached
+	// value or the cached error but don't ask topo again.
+	if !shouldRefresh {
+		if cacheValid {
+			return entry.value, nil
+		}
+		return nil, entry.lastError
+	}
+
+	// Refresh the state in a background goroutine if no refresh is already
+	// in progress none is already running. This way queries are not blocked
+	// while the cache is still valid but past the refresh time, and avoids
+	// calling out to the topo service while the lock is held.
+	if entry.refreshingChan == nil {
+		entry.refreshingChan = make(chan struct{})
+		entry.lastQueryTime = time.Now()
+
+		go func() {
+			defer func() {
+				if err := recover(); err != nil {
+					log.Errorf("ResilientQuery uncaught panic, cell :%v, err :%v)", key, err)
+				}
+			}()
+
+			newCtx, cancel := context.WithTimeout(ctx, *srvTopoTimeout)
+			defer cancel()
+
+			result, err := q.query(newCtx, entry)
+
+			entry.mutex.Lock()
+			defer func() {
+				close(entry.refreshingChan)
+				entry.refreshingChan = nil
+				entry.mutex.Unlock()
+			}()
+
+			if err == nil {
+				// save the value we got and the current time in the cache
+				entry.insertionTime = time.Now()
+				// Avoid a tiny race if TTL == refresh time (the default)
+				entry.lastQueryTime = entry.insertionTime
+				entry.value = result
+			} else {
+				q.counts.Add(errorCategory, 1)
+				if entry.insertionTime.IsZero() {
+					log.Errorf("ResilientQuery(%v, %v) failed: %v (no cached value, caching and returning error)", ctx, wkey, err)
+				} else if newCtx.Err() == context.DeadlineExceeded {
+					log.Errorf("ResilientQuery(%v, %v) failed: %v (request timeout), (keeping cached value: %v)", ctx, wkey, err, entry.value)
+				} else if entry.value != nil && time.Since(entry.insertionTime) < q.cacheTTL {
+					q.counts.Add(cachedCategory, 1)
+					log.Warningf("ResilientQuery(%v, %v) failed: %v (keeping cached value: %v)", ctx, wkey, err, entry.value)
+				} else {
+					log.Errorf("ResilientQuery(%v, %v) failed: %v (cached value expired)", ctx, wkey, err)
+					entry.insertionTime = time.Time{}
+					entry.value = nil
+				}
+			}
+
+			entry.lastError = err
+			entry.lastErrorCtx = newCtx
+		}()
+	}
+
+	// If the cached entry is still valid then use it, otherwise wait
+	// for the refresh attempt to complete to get a more up to date
+	// response.
+	//
+	// In the event that the topo service is slow or unresponsive either
+	// on the initial fetch or if the cache TTL expires, then several
+	// requests could be blocked on refreshingCond waiting for the response
+	// to come back.
+	if cacheValid {
+		return entry.value, nil
+	}
```

Comment:
This could very well be done before the check for refreshingChan.
Also refreshingCond -> refreshingChan in the comment.

Reply:
No, this is properly placed here. If `shouldRefresh` is true, we need to launch the background refresh goroutine _even if the cached value is freshe enough to return_, so this check needs to happen after.

Reply:
ah, makes sense.